internal/newusers/defs.go

```diff
@@ -0,0 +1,87 @@
+package newusers
+
+import (
+	"github.com/ubuntu/authd/internal/newusers/cache"
+)
+
+// UserInfo is the user information returned by the broker.
+type UserInfo struct {
+	Name  string
+	UID   int
+	Gecos string
+	Dir   string
+	Shell string
+
+	Groups []GroupInfo
+}
+
+// GroupInfo is the group information returned by the broker.
+type GroupInfo struct {
+	Name string
+	GID  *int
+}
+
+// UserEntry is the user information sent to the NSS service.
+type UserEntry struct {
+	Name  string
+	UID   int
+	GID   int
+	Gecos string
+	Dir   string
+	Shell string
+}
+
+// ShadowEntry is the shadow information sent to the NSS service.
+type ShadowEntry struct {
+	Name           string
+	LastPwdChange  int
+	MaxPwdAge      int
+	PwdWarnPeriod  int
+	PwdInactivity  int
+	MinPwdAge      int
+	ExpirationDate int
+}
+
+// GroupEntry is the group information sent to the NSS service.
+type GroupEntry struct {
+	Name  string
+	GID   int
+	Users []string
+}
+
+// userEntryFromUserDB returns a UserEntry from a UserDB.
+func userEntryFromUserDB(u cache.UserDB) UserEntry {
+	return UserEntry{
+		Name:  u.Name,
+		UID:   u.UID,
+		GID:   u.GID,
+		Gecos: u.Gecos,
+		Dir:   u.Dir,
+		Shell: u.Shell,
+	}
+}
+
+// shadowEntryFromUserDB returns a ShadowEntry from a UserDB.
+func shadowEntryFromUserDB(u cache.UserDB) ShadowEntry {
+	return ShadowEntry{
+		Name:           u.Name,
+		LastPwdChange:  u.LastPwdChange,
+		MaxPwdAge:      u.MaxPwdAge,
+		PwdWarnPeriod:  u.PwdWarnPeriod,
+		PwdInactivity:  u.PwdInactivity,
+		MinPwdAge:      u.MinPwdAge,
+		ExpirationDate: u.ExpirationDate,
+	}
+}
+
+// groupEntryFromGroupDB returns a GroupEntry from a GroupDB.
+func groupEntryFromGroupDB(g cache.GroupDB) GroupEntry {
+	return GroupEntry{
+		Name:  g.Name,
+		GID:   g.GID,
+		Users: g.Users,
+	}
+}
+
+// ErrCacheNoDataFound is the error returned when no entry is found in the cache.
+type ErrCacheNoDataFound = cache.NoDataFoundError
```

Comment:
It’s then `ErrNoDataFound`? No need to leak this is from the cache :)

Reply:
Well, then we should switch to wrapping the `ErrNeedsCleaning` as well, no? Otherwise, we are leaking that the cache is corrupted as well.

Reply:
Sorry if I am confused, but ErrNeedsCleaning will not be exported from users/? This is internal, isn’t it?

Reply:
It's "private" yes, but we don't remove it from the error chain when we "evaluate" the error. So, in the end, we might see an error message like: `database needs to be cleared and rebuilt: could not fetch something something`

Reply:
I think that’s fine, if people want to assert on it, it’s on them then, but let’s not reexport it and only let the text leaking through.